src/librustdoc/html/static/js/search.js

```diff
@@ -1267,44 +1223,98 @@ function initSearch(rawSearchIndex) {
              * @type Map<integer, FunctionType[]>
              */
             const fnTypeSet = new Map();
-            const addFnTypeToFnTypeSet = function addFnTypeToFnTypeSet(fnType) {
-                // Pure generic, or an item that's not matched by any query elems.
-                // Try [unboxing] it.
-                //
-                // [unboxing]:
-                // http://ndmitchell.com/downloads/slides-hoogle_fast_type_searching-09_aug_2008.pdf
-                const queryContainsArrayOrSliceElem = queryElemSet.has(typeNameIdOfArrayOrSlice);
-                if (fnType.id === -1 || !(
-                    queryElemSet.has(fnType.id) ||
-                    (fnType.id === typeNameIdOfSlice && queryContainsArrayOrSliceElem) ||
-                    (fnType.id === typeNameIdOfArray && queryContainsArrayOrSliceElem)
-                )) {
+            /**
+             * If `fnType` is a concrete type with generics, replace it with its generics.
+             *
+             * If `fnType` is a generic type parameter, replace it with any traits that it's
+             * constrained by. This converts `T where T: Trait` into the same representation
+             * that's directly used by return-position `impl Trait` and by `dyn Trait` in
+             * all positions, so you can directly write `-> Trait` and get all three.
+             *
+             * This seems to correspond to two transforms described in
+             * http://ndmitchell.com/downloads/slides-hoogle_fast_type_searching-09_aug_2008.pdf
+             * One of them is actual unboxing (turning `Maybe a` into `a`), while the other is
+             * equivalent to converting a type parameter into an existential, which Hoogle doesn't
+             * seem to actually do? I notice that these two searches produce the same result:
+             * https://hoogle.haskell.org/?hoogle=Eq+-%3E+Set+Eq+-%3E+Set+Eq
+             * https://hoogle.haskell.org/?hoogle=Ord+-%3E+Set+Ord+-%3E+Set+Ord
+             *
+             * Haskell is a bit of a foreign language to me. I just want to be able to look up
+             * Option combinators without having to actually remember their (mostly arbitrary)
+             * names, and Haskell claims to already have that problem solved...
+             *
+             * @type Map<integer, FunctionType[]>
+             */
+            const unbox = function unbox(fnType) {
```

Comment:
No need for a named function here.

```suggestion
            const unbox = function (fnType) {
```